### PR TITLE
issue  414 - Asynchronous task processing

### PR DESCRIPTION
**Pull request by abdullai-t**
Asynchronous task processing.

The following  was done:
1. Crud endpoints for super admins to create, update and delete tasks
2. A retry functionality in cases of task failure.

A use case for this could be sending quarterly reports via email to users (quarterly tasks). Or notifying users when policy or terms and conditions change (on-off task).

### CODE REVIEW

File: src/_main_/settings.py [hunk 1]
```diff
@@ -25,14 +25,14 @@
...
-RUN_SERVER_LOCALLY = False
+RUN_SERVER_LOCALLY = True
 if DJANGO_ENV == "local":
     RUN_SERVER_LOCALLY = True
 
 # Database selection, development DB unless one of these chosen
 IS_PROD = False
 IS_CANARY = False
-IS_LOCAL = False
+IS_LOCAL = True
```
Comment (Opoku-Agyemang) [inline src/_main_/settings.py:35]: intentional?

File: src/task_queue/migrations/0002_auto_20220315_2115.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+    dependencies = [
+        ('task_queue', '0001_initial'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='task',
+            name='recurring_interval',
+            field=models.CharField(blank=True, choices=[('EVERY_MINUTE', 'EVERY_MINUTE'), ('EVERY_HOUR', 'EVERY_HOUR'), ('EVERY_DAY', 'EVERY_DAY'), ('EVERY_MONTH', 'EVERY_MONTH'), ('EVERY_QUARTER', 'EVERY_QUARTER'), ('EVERY_YEAR', 'EVERY_YEAR'), ('ONE_OFF', 'ONE_OFF')], max_length=100, null=True),
```
[the remark this answers is not in the document]
Comment (abdullai-t) [reply src/task_queue/migrations/0002_auto_20220315_2115.py:16]: That is in a migration file.

File: src/_main_/settings.py (hunk 1 above)
Comment (abdullai-t) [reply src/_main_/settings.py:35]: will revert that

File: src/task_queue/migrations/0002_auto_20220315_2115.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sopokuagyemang) [reply src/task_queue/migrations/0002_auto_20220315_2115.py:16]: You're right

**Review by sopokuagyemang — APPROVED** [review]

File: src/_main_/settings.py [hunk 3]
```diff
@@ -286,3 +290,16 @@
 STATICFILES_LOCATION = 'static'
 MEDIAFILES_LOCATION = 'media'
 
+
+# Celery Setup
+CELERY_BROKER_URL = os.environ.get("CELERY_BROKER_URL")
```
Comment (sopokuagyemang) [inline src/_main_/settings.py:295]: How do we create one of these?

Comment (abdullai-t) [reply src/_main_/settings.py:295]: First of all, these packages have to be installed:
- redis
- celery
- django-celery-beat
- django-celery-results

After installing run `brew services start redis ` to start the Redis server (celery broker).
Add the redis server url to your env file like so
`CELERY_BROKER_URL = redis url` 
eg. ` CELERY_BROKER_URL='redis://localhost:6379/0'`

**Review by Opoku-Agyemang — APPROVED** [review]

**Pull request merged**